CHK-137: Google resolver for LocationSearch

wdsrocha opened the pull request:
#### What problem is this solving?

As title says, but check the [JIRA ticket](https://vtex-dev.atlassian.net/browse/CHK-137?atlOrigin=eyJpIjoiZjQxMmY5ZTQ1MWFjNDlhZTg0NTM1NWMxYzM0MmFlNzciLCJwIjoiaiJ9) for more details.

#### How should this be manually tested?

On [Graphiql](https://google--checkoutio.myvtex.com/_v/private/vtex.place-components@0.9.0/graphiql/v1), test `suggestAddresses` and `getAddressByExternalId` queries.

#### Notes

Please, read the commit messages when reviewing the changes, I tried to describe few things there.

wdsrocha commented on node/clients/GoogleGeolocation.ts:
```diff
@@ -6,10 +6,18 @@ export class GoogleGeolocation extends ExternalClient {
+
+  public suggestAddresses = (searchTerm: string, apiKey: string) => {
+    const input = encodeURI(searchTerm)
     return this.http.get(
-      `/geocode/json?latlng=${lat},${lng}&key=${apiKey}`,
+      `/place/autocomplete/json?input=${input}&key=${apiKey}`,
       {
-        metric: 'google-reverse-geocode-country',
+        metric: 'google-suggest-addresses',
         timeout: 3000,
       }
     )
```
This is not being used and should be removed (unless the `@googlemaps/google-maps-services-js` client get discarded)

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+    const response = await client.placeDetails({
+      params: {
+        place_id: id,
+        language: Language.pt_BR,
```
shouldn't we use the user's locale here?

wdsrocha replied:
@lucasecdb how do I get this info?

lucasecdb replied:
It is in `ctx.vtex.locale`

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+      process.env.VTEX_APP_ID
+    )
+
+    const client = new Client({})
```
this should be a client, much like the `GoogleGeolocation` client we have

wdsrocha replied:
How to do that? 🤔

wdsrocha replied:
(solved, just had to wrap the Google Client with a fake ExternalClient)

wdsrocha commented on graphql/types/AddressSuggestion.graphql:
```diff
@@ -0,0 +1,13 @@
+  mainTextMatchInterval: MatchInterval!
+  secondaryText: String!
+  externalAddressId: String
+  address: Address
```
Verificar se outros resolvers trazem essa informação (por @darlenedms)

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+    const rules = countryRules[country!]
+
+    if (!rules) {
+      console.warn(
```
Use `ctx.logger.warn` so we can get this warnings logged to Splunk

wdsrocha replied:
hmm, all there is in the `ctx` is `metricsLogger`, which have a `add` and `getSummaries` methods. Do you know if the equivalent of what you suggested would be `ctx.metricsLogger.add(`?

lucasecdb replied:
oh, sorry! it is `ctx.vtex.logger` if I'm not mistaken

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+
+    const geolocationRules = rules.geolocation
+
+    function setAddressFields() {
```
can we move this function outside the resolver body?

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+      )
+    }
+
+    function setGeoCoordinates(updatedAddress: Address) {
```
same as above

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+      return updatedAddress
+    }
+
+    function setCountry(updatedAddress: Address) {
```
same as above

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+      return updatedAddress
+    }
+
+    return pipe(
```
can we _not_ use ramda? 😅 we are kinda moving away from using it and prefer using native JS functions (although in this case it doesn't exist yet)

lucasecdb commented on node/resolvers/search.ts:
```diff
@@ -0,0 +1,214 @@
+      process.env.VTEX_APP_ID
+    )
+
+    const client = new Client({})
```
same as the other comment about the client

lucasecdb commented on node/clients/Google.ts:
```diff
@@ -0,0 +1,12 @@
+import { ExternalClient, IOContext, InstanceOptions } from '@vtex/api'
+
+export class Google extends ExternalClient {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  externalClient: Client
```
you could add the methods to this class that would just proxy them to this `externalClient` in order to avoid exposing it directly to the resolvers, what do you think?

wdsrocha replied:
Why do you think exposing the external client would be bad?

lucasecdb replied:
for a common API surface. think if we ever want to change the google maps library (don't know why we would though), we wouldn't need to change the calls on every resolver, just on the client. anyway, this was more like a question 😅 don't feel the need to do this rn

wdsrocha replied:
I see. Dw, let's go for it (it was really quick to do and it may pay off in the future, so why not 👀)

wdsrocha committed:
Add queries for autocomplete service

The rational behind having for two queries for autocomplete is that most APIs don't provide a complete address/place object for the autocomplete route, so N+1 requests would be necessary to get all addresses required for the suggestAddress query.

By making having a `externalId` attribute, the `suggestAddresses` query will just return a special id that will be followed up by a lookup query on the selected address. Basically, the flow for google would be:

```
*User starts typing*

                  suggestAddresses(searchTerm)
place-components -----------------------------> places-graphql
place-components <----------------------------- places-graphql
                  AddressSuggestion[] where
				  each element has an externalId

*User selects a suggestion*

                   getAddressByExternalId(id)
place-components ----------------------------> places-graphql
place-components <---------------------------- places-graphql
                            Address

And this returned address will be set on the Context
```

wdsrocha committed:
Add new 'robust' client for Google

Makes development easier and looks light:
https://bundlephobia.com/result?p=@googlemaps/google-maps-services-js@3.1.6

wdsrocha committed:
Add search resolver for autocomplete queries

Based on the `geolocation.ts`. Some refactoring were made and it would be cool to make those same refactors on the original `geolocation.ts` file.

wdsrocha committed:
Format GoogleGeolocation.ts

wdsrocha committed:
Update CHANGELOG

alvim reviewed (approved):
Nice! LGTM.
Your code looks really clean and your dedication to commenting on commits in detail is remarkable.

As a tip for the future I would just suggest you to be more detailed in manual test description. For example, I think that someone who has no experience with GraphQL would not be able to test with that description.

wdsrocha commented:
Thanks for the feedback, @alvimm !
